Source/Renderer/modernizeShader.js

```diff
@@ -0,0 +1,233 @@
+            var line = splitSource[number];
+            splitSource[number] = line.replace(regex, '$1' + replacement + '$3');
+        }
+    }
+
+    function replaceInSourceRegex(regex, replacement, splitSource) {
+        for (var number = 0; number < splitSource.length; ++number) {
+            var line = splitSource[number];
+            splitSource[number] = line.replace(regex, replacement);
+        }
+    }
+
+    function findInSource(str, splitSource) {
+        var regexStr = '(^|[^\\w])(' + str + ')($|[^\\w])';
+        var regex = new RegExp(regexStr, 'g');
+        for (var number = 0; number < splitSource.length; ++number) {
+            var line = splitSource[number];
+            if (regex.test(line)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    function compileSource(splitSource) {
+        var wholeSource = '';
+        for (var number = 0; number < splitSource.length; ++number) {
+            wholeSource += splitSource[number] + '\n';
+        }
+        return wholeSource;
+    }
+
+    function setAdd(variable, set1) {
+        if (set1.indexOf(variable) === -1) {
+            set1.push(variable);
+        }
+    }
+
+    function getVariablePreprocessorBranch(layoutVariables, splitSource) {
+        var variableMap = {};
+
+        var numLayoutVariables = layoutVariables.length;
+        for (var a = 0; a < numLayoutVariables; ++a) {
+            var variableThatWeCareAbout = layoutVariables[a];
+            variableMap[variableThatWeCareAbout] = [null];
+        }
+
+        var stack = [];
+        for (var i = 0; i < splitSource.length; ++i) {
+            var line = splitSource[i];
+            var hasIF = /(#ifdef|#if)/g.test(line);
+            var hasELSE = /#else/g.test(line);
+            var hasENDIF = /#endif/g.test(line);
+
+            if (hasIF) {
+                stack.push(line);
+            } else if (hasELSE) {
+                var top = stack[stack.length - 1];
+                var op = top.replace('ifdef', 'ifndef');
+                if (/if/g.test(op)) {
+                    op = op.replace(/(#if\s+)(\S*)([^]*)/, '$1!($2)$3');
+                }
+                stack.pop();
+                stack.push(op);
+            } else if (hasENDIF) {
+                stack.pop();
+            } else if (!/layout/g.test(line)) {
+                for (var varIndex = 0; varIndex < numLayoutVariables; ++varIndex) {
+                    var varName = layoutVariables[varIndex];
+                    if (line.indexOf(varName) !== -1) {
+                        if (variableMap[varName].length === 1 && variableMap[varName][0] === null) {
+                            variableMap[varName] = stack.slice();
+                        } else {
+                            variableMap[varName] = variableMap[varName].filter(function(x) {
+                                return stack.indexOf(x) >= 0;
+                            });
```

Comment:
I'm still confused by this areas and the use of `null`. Could you explain what is going on?